printing: Attempt at speeding up ccode(Min(*long_list))

clyring opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #15424.


#### Brief description of what is fixed or changed
_print_math_func in sympy.printing.ccode.C99CodePrinter no longer recursively prints sub-objects when run with `nest=True`, making the printing of Min objects with many arguments much faster.

#### Other comments
Only the printing of the Min objects is made faster: Constructing a Min object from a long argument list remains slow.

#### Release Notes
<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* printing
  * printing C code for Min and Max objects with many arguments is much faster
<!-- END RELEASE NOTES -->

clyring committed:
printing: Attempt at speeding up ccode(Min(*long_list))

asmeurer commented:
Can we add a test for something that would be too slow without this, or is it too slow to construct the Min?

clyring commented:
I confess I'm not sure what the criteria for such a test are. Currently on my machine it takes about 3 seconds to construct a 50-arg Min object. Subsequently generating the C code takes about 45 seconds on master and about a millisecond on the topic branch.

asmeurer commented:
Such a test should be fine. We should also figure out how to make the Min construction faster. It seems even `evaluate=False` doesn't speed it up.

asmeurer commented:
It seems the problem is in `_find_localzeros`, which has quadratic runtime. It's also not skipped when using `evaluate=False`.

clyring committed:
Reworked C89 Max/Min printing, added performance test

The C89 code generated for Max and Min objects with n arguments
now has quadratic rather than exponential length. Minor additional
cleanup was done in the analagous C99 code printing.

oscargus commented:
Is anything else required here?

oscarbenjamin commented:
I'm not sure why I added the author's turn label. I've just removed it.

asmeurer commented:
The test I requested was added. We should still make Min itself faster as I described, but that's a separate issue.

asmeurer commented:
Opened https://github.com/sympy/sympy/issues/16249 on the Min constructor performance.
